src/coreclr/vm/interoplibinterface_java.cpp

```diff
@@ -0,0 +1,193 @@
+                continue;
+
+            MarkCrossReferenceArgs* mcrargs = g_mcrargs;
+            InterlockedExchange((LONG*)&g_mcrargs, NULL);
+
+            g_MarkCrossReferences(mcrargs->sccsLen, mcrargs->sccs, mcrargs->ccrsLen, mcrargs->ccrs);
+
+            FreeArgs(mcrargs);
+        }
+
+        return 0;
+    }
+}
+
+extern "C" BOOL QCALLTYPE JavaMarshal_Initialize(
+    _In_ void* markCrossReferences)
+{
+    QCALL_CONTRACT;
+    _ASSERTE(markCrossReferences != NULL);
+
+    BOOL success = FALSE;
+
+    BEGIN_QCALL;
+
+    // Switch to Cooperative mode since we are setting callbacks that
+    // will be used during a GC and we want to ensure a GC isn't occurring
+    // while they are being set.
+    {
+        GCX_COOP();
+        if (InterlockedCompareExchange((LONG*)&g_Initialized, TRUE, FALSE) == FALSE)
+        {
+            g_MarkCrossReferences = (CrossreferenceHandleCallback)markCrossReferences;
+
+            g_BridgeReady = FALSE;
+            g_BridgeTrigger = new CLREvent();
+            g_BridgeTrigger->CreateAutoEvent(FALSE);
+
+            g_BridgeThread = SetupUnstartedThread();
+            BOOL createdThread = g_BridgeThread->CreateNewThread(0, &BridgeThreadStart, NULL, W("JavaBridge"));
+            _ASSERTE(createdThread != FALSE);
+            g_BridgeThread->StartThread();
+            success = TRUE;
+        }
+    }
+
+    END_QCALL;
+
+    return success;
+}
+
+extern "C" void* QCALLTYPE JavaMarshal_CreateReferenceTrackingHandle(
+    _In_ QCall::ObjectHandleOnStack obj,
+    _In_ void* context)
+{
+    QCALL_CONTRACT;
+
+    OBJECTHANDLE instHandle = NULL;
+
+    BEGIN_QCALL;
+
+    GCX_COOP();
+    instHandle = GetAppDomain()->CreateCrossReferenceHandle(obj.Get(), context);
+
+    END_QCALL;
+
+    return (void*)instHandle;
+}
+
+extern "C" BOOL QCALLTYPE JavaMarshal_GetContext(
+    _In_ OBJECTHANDLE handle,
+    _Out_ void** context)
+{
+    QCALL_CONTRACT;
+    _ASSERTE(handle != NULL);
+    _ASSERTE(context != NULL);
+
+    BOOL success = FALSE;
+
+    BEGIN_QCALL;
+
+    IGCHandleManager* mgr = GCHandleUtilities::GetGCHandleManager();
+    HandleType type = mgr->HandleFetchType(handle);
+    if (type == HNDTYPE_CROSSREFERENCE)
+    {
+        *context = mgr->GetExtraInfoFromHandle(handle);
+        success = TRUE;
+    }
+
+    END_QCALL;
+
+    return success;
+}
+
+void JavaNative::TriggerGCBridge(
+    _In_ size_t sccsLen,
+    _In_ StronglyConnectedComponent* sccs,
+    _In_ size_t ccrsLen,
+    _In_ ComponentCrossReference* ccrs)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
+
+    if (g_MarkCrossReferences == NULL)
+        return;
+
+    MarkCrossReferenceArgs* mcrargs = new MarkCrossReferenceArgs();
+    mcrargs->sccsLen = sccsLen;
+    mcrargs->sccs = sccs;
+    mcrargs->ccrsLen = ccrsLen;
+    mcrargs->ccrs = ccrs;
+
+    if (g_BridgeReady == FALSE)
+    {
+        FreeArgs(mcrargs);
+        return;
+    }
+
+    g_mcrargs = mcrargs;
```

Comment:
There seems to be a subtle race condition. If the system is overloaded and it did not get a chance to run since the last turn of the crank, we can still see `g_BridgeReady` set to `TRUE`, but `g_mcrargs` is still going to point to the old payload that we are going to leak by overwriting the point.

Reply:
This entire thread operation is completely dirty and a simple prototype. It all needs to be redone.